kolibri/core/device/serializers.py

```diff
@@ -56,6 +59,18 @@ class Meta:
             "allow_guest_access",
         )
 
+    def validate(self, data):
+        if (
+            "superuser" not in data
+            and GET_USERNAME in interface
+            and "request" in self.context
+            and valid_app_key_on_request(self.context["request"])
+        ):
+            data["os_user"] = True
```

Comment:
One issue I have with this approach is our authentication becomes "knows the app key." The trouble is, every client application needs that same key in order to work, but we don't necessarily want every client application to be able to authenticate as any user.

In the implementation of this we have with Kolibri GNOME + Endless OS (https://github.com/endlessm/kolibri-desktop-auth-plugin/), we ended up using a separate authentication token for user logins. That's because, on Endless OS, we have a system-wide Kolibri service that multiple desktop users may talk to. So, each of those users is allowed to see the app key. Instead, we need a different key to be able to say "I am _this_ user."

Reply:
So, the issue is that the service is running as a global user, so getting the OS Username in the backend will always give that user, and not the user that is actually running the app?

So, we could embellish the OSUser model to also generate a unique auth token here - and that gets returned as part of the `get_initialization_url` call, if the capability for getting the OSUser is registered. That means we would move the dynamic user creation from the initialize API endpoint, to the get_initialize_url method, and then add the auth token as an extra (optional) parameter in the initialization URL.

What we would then need to do for the device provision serializer, is instead provide the OSUsername in the provisioning request, in order to generate the user appropriate, as the backend might be running in a different OSUser context.

Does that all seem correct?

Reply:
> So, the issue is that the service is running as a global user, so getting the OS Username in the backend will always give that user, and not the user that is actually running the app?

Indeed! And also it's a long-running service that may be used by multiple desktop users in the same session.

> So, we could embellish the OSUser model to also generate a unique auth token here - and that gets returned as part of the `get_initialization_url` call, if the capability for getting the OSUser is registered. That means we would move the dynamic user creation from the initialize API endpoint, to the get_initialize_url method, and then add the auth token as an extra (optional) parameter in the initialization URL.
>
> What we would then need to do for the device provision serializer, is instead provide the OSUsername in the provisioning request, in order to generate the user appropriate, as the backend might be running in a different OSUser context.
> 
> Does that all seem correct?

Okay, so the desktop app would get back its own auth token and hold on to that to authenticate with? I think that makes sense. We'd lose a bit of verifiability. In particular, we wouldn't be able to say whether the user is who they claim to be, or an administrator or a standard user. But we'd still get some pleasant behaviour from the app itself.

The moving part in all this is we need something that can reliably say "this client is this desktop user," which is challenging with HTTP. I think that's the spot where we need the ability to plug in a more fine tuned solution. A good tool for that on Linux is D-Bus. So, a client asks a trusted D-Bus service for a token, and that service generates a token and remembers which user it belongs to. (It can also talk to another service to get some details about that user). Then our Kolibri plugin talks to the same D-Bus service and asks it "who is this token for?". But of course, on other platforms we'd have to do that a different way entirely.

Reply:
The point of the initialization_url is to have a shared secret between the Kolibri backend and the webview that the app frontend is using, what I am suggesting is that we enhance that to be user specific. Rather than just 'this is an app controlled webview', it's now 'this is a webview controlled by this user'.

I note that you're not using the app interface anywhere in the GNOME app in order to provide this sort of guarantee, but it seems like the modules in there do have programmatic access to Kolibri, so this should be workable?

The code that is responsible for loading the webview would therefore be responsible for getting a URL that encodes both the app-key shibboleth, and the new Kolibri generated auth token I am proposing. This would then allow the backend to login to the appropriate user before redirecting to the page that you actually want to go to.

The capability I have defined here for 'get_user' could in theory call out to the D-Bus to get the correctly authenticated username, and ensure that it never returns the wrong one, so this system is already pluggable to the extent that I think is necessary - moving the tokens external to Kolibri seems like extra complexity.

This does not resolve whether the user is a superadmin or not - we could expand the app interface capabilities to get 'this user is an admin' to determine Kolibri permissions as well.

Reply:
Just to check my understanding of what's at stake here, by listing out some of my assumptions/interpretations (to be corrected if off-base, plz!):
- This seems specifically relevant for the case where there's a system-wide installation of the service with a shared cross-user KOLIBRI_HOME, which is currently only the case for the GNOME app.
- The concern @dylanmccall is raising is that if the frontend app (which runs in user-space) is able to choose which user to log in as, without some way to prove to the backend that it should have the rights to that user, then someone logged in as user A could (with some poking around with internals, but within the rights of their OS user account) log in and access the data of another user.
- This concern is specifically relevant if we're *not* making everyone a super admin, as noted, since otherwise they can access any user's data anyway once logged in.
- The concern is also only relevant if the OS user accounts can't directly read from the KOLIBRI_HOME directory, since otherwise they could read from the Kolibri DB and access any user's data as well. I imagine the app could be made to function without access to KOLIBRI_HOME except from the background system service, but I'm not sure of whether it currently does that.
- All that being said, if the prior two bullets do hold up, I do understand the concern in principle, and agree that it would be ideal to restrict access for cases like this (i.e. specifically for the way the GNOME app behaves, at least on Endless OS -- I don't think it would currently apply anywhere else?).
- The reason @dylanmccall proposes using D-Bus as a way to get a user-specific token is that it's one way for the app to demonstrate to the Kolibri server process that the token request is coming from a particular actual user, rather than taking it at its word.

Regarding that last bullet, another alternative that comes to mind would be creating and storing a token (non-world-readable) in the user's own data directory, and then associating that token with the user in the backend -- and then the app could re-use that token for re-logging in.[1] For creating a super-admin account, it could use something like visudo to drop its token into a "known admin tokens" directory in KOLIBRI_HOME, thereby proving that the user of that token should be allowed to have an admin account in Kolibri. Not sure whether that's more, or less, general-purpose than relying on D-Bus, but might achieve similar ends.

[1] Update: after reading back through the thread again, I think this is basically where the conversation was heading -- with "the desktop app would get back its own auth token and hold on to that to authenticate with". While the simple version of that approach covers the "can only log in as that same user when you return" case, the piece needs to be covered is distinguishing whether to make the OSUser an admin or learner -- which it seems could be addressed either by D-Bus or the `visudo` type approach I described above.

Reply:
Yeah, I think that would work nicely. It's pretty ordinary for a web service to have a one-time login token system anyway, so plugging in to that sort of thing from get_initialize_url seems pretty reasonable. It also helps to simplify the setup flow, which would be great :)

For a system where we have less trust between the server and the app (such as when the server is a d-bus daemon running as another user), we can have another API that calls `get_initialize_url` with stuff like username filled in automatically.

Reply:
>In order to accommodate the provisioning case, if the device is not provisioned, we would need to set the token as a temporary cookie in the initialization endpoint, so that the device provisioning can read that to know to create a superuser for that token

If I'm understanding correctly: If the desired behavior is that the first user to interact with Kolibri becomes the sole superuser, then this makes sense. The other direction being discussed above was having a user's superuser status be determined by whether they OS-level admin access (e.g. being root, or a sudoer).

Reply:
I think I've addressed all the above concerns - to handle the provisioning case, I instead avoid creating a super user during provisioning, and leave it up to the app to redirect the user to the initialize URL after provisioning to create a user or superuser, depending on the returned value from the `get_os_user` capability, which is now expected to return a tuple of `username, is_superuser`.

Let me know what you think!